File: src/main/java/leetcode/dfs/JumpGame3.java
package leetcode.dfs;

public class JumpGame3 {

    public boolean canReach(int[] arr, int idx) {
        boolean[] visited = new boolean[arr.length];
        return canReach(arr, visited, idx);
    }

    public boolean canReach(int[] arr, boolean[] visited, int idx) {

        if (idx < 0 || idx >= arr.length || visited[idx])
            return false;

        if (arr[idx] == 0)
            return true;

        visited[idx] = true;

        return canReach(arr, visited, idx + arr[idx]) || canReach(arr, visited, idx - arr[idx]);
    }
}

File: src/main/python/leetcode/array/jump_game_3.py
from typings import List, Set

class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:

        return self.canReachInternal(arr, set(), start)


    def canReachInternal(self, arr: List[int], visited: Set[int], start: int) -> bool:

        if start < 0 or start >= len(arr) or (start in visited):
            return False

        if arr[start] == 0:
            return True

        visited.add(start)

        return self.canReachInternal(arr, visited, start+arr[start]) or self.canReachInternal(arr, visited, start-arr[start])